[193] Dataset update checks

Jack-0-0 opened the pull request:
Closes #193.

This PR:
* Updates `innovation_sweet_spots/getters/gtr_2022.py` to use just one function that can load csv or json files.
* Adds a script to check new versions of datasets for Crunchbase, GtR and NIHR (`tests/data_update_checks/data_update_checks.py`)
* Functions used in `data_update_checks.py` across all three datasets can be found at `tests/data_update_checks/utils.py`
* Dataset specific functions used in `data_update_checks.py` can be found at:
  * `tests/data_update_checks/cb/cb_utils.py`
  * `tests/data_update_checks/gtr/gtr_utils.py`
  * `tests/data_update_checks/nihr/nihr_utils.py`

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

Jack-0-0 committed:
Add data update checks file and utils

Jack-0-0 committed:
Update gtr 2022 getters

Jack-0-0 committed:
Add dataset specific plots to data_update_checks

Jack-0-0 committed:
Update utils logging messages and to specify dataset versions

Jack-0-0 committed:
Add cb_utils

Jack-0-0 committed:
Add gtr_utils

Jack-0-0 committed:
Add nihr_utils

Jack-0-0 committed:
Remove commented checks

Jack-0-0 committed:
Add missing typehints

Jack-0-0 committed:
Update docstrings

Jack-0-0 committed:
Add requirements

Jack-0-0 committed:
Add readme

Jack-0-0 committed:
Update readme

Jack-0-0 committed:
Update readme

Jack-0-0 committed:
Add exception handling

Jack-0-0 committed:
Uncomment line

Jack-0-0 committed:
Update plot dirs to have new and current data version names

Jack-0-0 commented on innovation_sweet_spots/tests/data_update_checks/gtr/gtr_utils.py:
```diff
@@ -0,0 +1,146 @@
+    """Group gtr funding by year and process into format
+    that can be used in the function `gtr_funding_per_year_plot`"""
+    return (
+        load_file_as_df(path, "gtr_projects-wrangled_project_data.csv")[
+            ["fund_start", "amount"]
+        ]
```
Do we want to use file `"gtr_projects-wrangled_project_data.csv"` here?

beingkk replied:
Yes, let's keep it like that.

Jack-0-0 commented on innovation_sweet_spots/tests/data_update_checks/gtr/gtr_utils.py:
```diff
@@ -0,0 +1,146 @@
+    """Group gtr count of projects by year and process into format
+    that can be used in the function `gtr_projects_per_year_plot`"""
+    return (
+        load_file_as_df(path, "gtr_projects-wrangled_project_data.csv")[
+            ["fund_start", "project_id"]
+        ]
```
Do we want to use file `"gtr_projects-wrangled_project_data.csv"` here?

beingkk replied:
Sure, let's leave it like it is.

beingkk replied:
Yes, let's keep it like that.

beingkk commented on innovation_sweet_spots/tests/data_update_checks/utils.py:
```diff
@@ -0,0 +1,193 @@
+        logger.info(f"{cur_dir}/ and {new_dir}/ files match.")
+    else:
+        logger.warning(f"{cur_dir}/ and {new_dir}/ files do not match.")
+        if missing_files := cur_files - new_files:
```
Hey, this line is throwing a syntax error for me:

```
Traceback (most recent call last):
  File "innovation_sweet_spots/tests/data_update_checks/data_update_checks.py", line 1, in <module>
    from innovation_sweet_spots.tests.data_update_checks.utils import (
  File "/Users/karliskanders/Documents/innovation_sweet_spots/innovation_sweet_spots/tests/data_update_checks/utils.py", line 37
    if missing_files := cur_files - new_files:
                      ^
SyntaxError: invalid syntax
```

beingkk replied:
Probably because I'm still in a Python 3.7 environment. I wonder if there is an equivalent syntax that still works for 3.7? (otherwise I can also upgrade my environment)

beingkk commented on innovation_sweet_spots/tests/data_update_checks/utils.py:
```diff
@@ -0,0 +1,193 @@
+    new_dir: str,
+):
+    """Check that the number of columns in the current and new dataframes are
+    the ame. Log the outcome and provide additional information on the
```
```suggestion
    the same. Log the outcome and provide additional information on the
```

beingkk commented on innovation_sweet_spots/tests/data_update_checks/utils.py:
```diff
@@ -0,0 +1,193 @@
+        if add_cols := new_cols - cur_cols:
+            logger.warning(
+                f"Compared with {cur_dir}/{filename}, {new_dir}/{filename} dataset has {len(add_cols)} additional columns: {add_cols}."
+            )
```
I wonder if it would be useful here to also list the names of the extra or missing columns?

beingkk commented on innovation_sweet_spots/tests/data_update_checks/utils.py:
```diff
@@ -0,0 +1,193 @@
+            logger.warning(
+                f"Compared with {cur_dir}/{filename}, {new_dir}/{filename} dataset is missing {len(missing_cols)} columns: {missing_cols}."
+            )
+        if add_cols := new_cols - cur_cols:
```
I guess 'add' in `add_cols` stands for 'additional' ?

Jack-0-0 replied:
Yes, that's correct.

beingkk commented on innovation_sweet_spots/tests/data_update_checks/README.md:
```diff
@@ -0,0 +1,17 @@
+
+## Usage
+
+Run: `python innovation_sweet_spots/tests/data_update_checks/data_update_checks.py [current_directory] [new_directory] [dataset]`
```
```suggestion
Run: `python innovation_sweet_spots/tests/data_update_checks/data_update_checks.py [old_version_directory] [new_version_directory] [dataset]`
```

beingkk commented on innovation_sweet_spots/tests/data_update_checks/README.md:
```diff
@@ -0,0 +1,17 @@
+# Dataset Update Checks
+
+`data_update_checks.py` compares the current version of a dataset with a new version. It checks for missing/additional files, rows, columns, NaNs. It also creates some dataset specific plots to see how the coverage for each year has changed.
```
```suggestion
`data_update_checks.py` compares two versions of a dataset (eg, and older version with a new version). It checks for missing/additional files, rows, columns, NaNs. It also creates some dataset specific plots to see how the coverage for each year has changed.
```

beingkk commented on innovation_sweet_spots/tests/data_update_checks/README.md:
```diff
@@ -0,0 +1,17 @@
+
+## Requirements
+
+In addition to the main Innovation Sweet Spots requirements, there are some additional requirements needed to run the dataset update checks. To install these, run: `pip install -r innovation_sweet_spots/tests/data_update_checks/requirements.txt`.
```
Hey @Jack-0-0 looks like there is only one requirement there - could we add it to the main requirements.txt file? (located at the root of the repo)

Jack-0-0 replied:
I've moved the requirement into the main `requirements.txt` file and removed mention of requirements from the readme.

beingkk commented on innovation_sweet_spots/tests/data_update_checks/utils.py:
```diff
@@ -0,0 +1,193 @@
+    logger.info(f"NaN plot saved to {save_path}")
+
+
+def run_data_checks(
```
Hey @Jack-0-0 , would it be possible to slightly generalise this function by allowing a case where the `dataset = None`, and the util could still compare the tables between both folders, but wouldn't run the dataset-specific tests?

beingkk commented on innovation_sweet_spots/tests/data_update_checks/utils.py:
```diff
@@ -0,0 +1,193 @@
+        palette="Paired",
+    ).set(
+        title=f"Percentage of NaNs in {cur_dir} vs. {new_dir} for {filename}",
+        xlim=(0, 1),
```
Hey, this limit is too tight for many of the plots - probably best to not set the max limit.

See these examples:

![crunchbase_investors_nan_plot](https://user-images.githubusercontent.com/29844253/200627996-0758b8be-da16-4e11-8a58-56cca8f041b6.png)

![crunchbase_organizations_nan_plot](https://user-images.githubusercontent.com/29844253/200628030-36756f1c-e868-4fa5-8555-002b166a06fa.png)

Jack-0-0 replied:
Good spot, thank you. I've removed `xlim` from the plots.

beingkk commented on innovation_sweet_spots/tests/data_update_checks/data_update_checks.py:
```diff
@@ -0,0 +1,33 @@
+DATA_PATH = PROJECT_DIR / "inputs/data"
+
+
+def run_update_checks(cur_dir: str, new_dir: str, dataset: str):
```
Hey @Jack-0-0 , would it be possible to slightly generalise this function by allowing a case where the `dataset = None`, and the function could still compare the tables between both folders, but wouldn't run the dataset-specific tests?

Jack-0-0 replied:
I've updated the `data_update_checks.py` so that you can now compare any two directories in `inputs/`.

beingkk reviewed (changes requested):
Hi Jack, thank you very much for this!

I tested this on python 3.9 and it ran very well - logs are very clear and useful, well done!

One issue I noticed is that the nan plots have an axis limit at 1% which cuts off many plots.

Besides that, I've also suggested if perhaps the main script can be slightly generalised (I've added comment in the code).

The rest is just very minor comments, and then I think it'll be good to go.

beingkk commented:
Oh - one last thing, @Jack-0-0  - could you please update gitignore to not include the png chart files? I don't think we need to upload them on GitHub - we could push them on s3 in the future, but for now let's just leave it.

Jack-0-0 committed:
Fix typo

Jack-0-0 committed:
Update readme usage example

Jack-0-0 committed:
Update readme intro

Jack-0-0 committed:
Move data update checks requirements to main requirements file

Jack-0-0 committed:
Remove xlim on plots

Jack-0-0 committed:
Allow run_update_checks to run without specifying dataset

Jack-0-0 committed:
Update docstring to note outcome when no dataset argument passed

Jack-0-0 committed:
Update gitignore to ignore data update check plots

Jack-0-0 commented:
> Oh - one last thing, @Jack-0-0 - could you please update gitignore to not include the png chart files? I don't think we need to upload them on GitHub - we could push them on s3 in the future, but for now let's just leave it.

I've updated the gitignore to ignore the .png files, see [here](https://github.com/nestauk/innovation_sweet_spots/pull/194/commits/d6ed0bcf636970d7ad4d422292eebe1b600965bb).

Jack-0-0 commented:
> Hi Jack, thank you very much for this!
> 
> I tested this on python 3.9 and it ran very well - logs are very clear and useful, well done!
> 
> One issue I noticed is that the nan plots have an axis limit at 1% which cuts off many plots.
> 
> Besides that, I've also suggested if perhaps the main script can be slightly generalised (I've added comment in the code).
> 
> The rest is just very minor comments, and then I think it'll be good to go.

Hi @beingkk. Thank you for the review, I've addressed your comments. Let me know if it's good to merge.

beingkk reviewed (approved):
Thank you, looks good to merge!
